### PR TITLE
Fix context propagation across goroutines

### PR DESCRIPTION
**Pull request by edeNFed**
## What's changed
1. Added 4 utility functions that should be used by instrumentors:
- `get_consistent_key`: This function provides a key that works across any version of Go. This key should be used in the eBPF map created by the instrumentor to record the end timestamp at the end of the probed function.
- `start_tracking_span(void *contextContext, struct span_context *sc)` - This function should be used in order to perform in-process context propagation, meaning passing context between different instrumentations in the same applications (see below test case for example).
- `stop_tracking_span(struct span_context *sc)` - This function marks the passed span context as no more relevant for context propagation. This function should usually be called when recording the end timestamp of a span.
- `get_parent_span_context(void *ctx)` - Use this function in order to get the parent span context if available. This function receives the address of `context.Context` object.

2. Modified existing instrumentors to use those functions when reading/writing to eBPF maps and to performing context propagation between instrumentors.

3. Modified gRPC (client) instrumentation to inject headers by probing `loopyWriter.headerHandler` instead of `http2Client createHeaderFields`: In Go >= 1.17, the return value is passed by registers and not by the stack. Currently, eBPF is not able to modify registers' values, only memory. Changing the uprobe to attach to `loopyWriter.headerHandler` allow us to add context-propagation headers even on Go versions above 1.17.

## Validation
I used the following code as a test case:
```go
func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		con, err := grpc.Dial("grpc-server.default:50051", grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			panic(err)
		}

		stopCh := make(chan struct{})
		go func() {
			healthClient := healthpb.NewHealthClient(con)
			resp, err := healthClient.Check(r.Context(), &healthpb.HealthCheckRequest{})
			if err != nil {
				panic(err)
			}

			fmt.Fprintf(w, "gRPC server status: %s", resp.Status)
			close(stopCh)
		}()

		<-stopCh
	})

	fmt.Println("HTTP server started on port 8080")
	err := http.ListenAndServe(":8080", nil)
	if err != nil {
		panic(err)
	}
}
```

Which produces the following trace (after the applied fix)
Screenshot 2023-05-22 at 1 11 33 PM

**Tested on Go 1.15 and 1.20, both on arm64 and amd64**

### CODE REVIEW

**Comment by edeNFed** [comment]
Blocked by: #82

**Comment by MrAlias** [comment]
Related to https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/86

**Comment by edeNFed** [comment]
@open-telemetry/go-instrumentation-approvers this is ready for review, thanks! 🙏

File: include/go_context.h
Comment (MikeGoldsmith) [inline include/go_context.h:72]: For consistency with `stop_tracking_span`, could we rename to `start_tracking_span`?

File: include/go_context.h
Comment (MikeGoldsmith) [inline include/go_context.h:75]:
```suggestion
static __always_inline void track_running_span(void *ctx, struct span_context *sc) {
    bpf_map_update_elem(&tracked_spans, &ctx, sc, BPF_ANY);
    bpf_map_update_elem(&tracked_spans_by_sc, sc, &ctx, BPF_ANY);
}
```

File: pkg/instrumentors/bpf/google/golang/org/grpc/bpf/probe.bpf.c
Comment (MikeGoldsmith) [inline pkg/instrumentors/bpf/google/golang/org/grpc/bpf/probe.bpf.c:156]: Is it intentional to monitor `loopyWriter_headerHandler` instead of `Http2Client_createHeaderFields`? Do we need to still monitor `http2Client_NewStream` too?

Can you say anymore about why it's better to use this instead?

**Review by MikeGoldsmith — COMMENTED** [review]
This looks really good, thanks @edeNFed.

I've left some comments and suggestions.

It's worth noting #91 conflicts with this. Whichever PR lands first, will require the other to be rebased.

File: pkg/instrumentors/bpf/google/golang/org/grpc/bpf/probe.bpf.c
Comment (edeNFed) [reply pkg/instrumentors/bpf/google/golang/org/grpc/bpf/probe.bpf.c:156]: Added description to PR

**Comment by edeNFed** [comment]
@open-telemetry/go-instrumentation-approvers This is ready for another review
Thank you 🙏

**Review by MrAlias — APPROVED** [review]

File: pkg/instrumentors/bpf/net/http/client/bpf/probe.bpf.c
Comment (MikeGoldsmith) [inline pkg/instrumentors/bpf/net/http/client/bpf/probe.bpf.c:56]: My understanding is `bpf_printk` has non-negligible performance impact. Maybe we should review where we use them, and have some sort of flag to enable / disable? There are a few 

I'll create a ticket to track from this comment.

**Review by MikeGoldsmith — APPROVED** [review]
Looks great, thanks @edeNFed 👍🏻

**Pull request merged**
